File: scripts/generate_schemas.py
import json
import os

# h/t: https://www.safaribooksonline.com/library/view/regular-expressions-cookbook/9781449327453/ch04s07.html
ISO_DATE_PATTERN = r'^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])(\.[0-9]+)?(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?$'

USERNAME_PATTERN = r'^(\*|[^\s\-\*]+\-\*|[^\s\-\*]+\-[^\s\-\*]+\-\*|[^\s\-\*]+\-[^\s\-\*]+\-[^\s\-\*]+)$'

JAM_ID_PATTERN = r'[\w]+\.[\w]+\.[\w]+'

ADMIN = {
    "type": "jsonschema",
    "schema": {
        "id": "admin",
        "type": "object",
        "properties": {
            "id": {
                "id": "id",
                "type": "string"
            },
            "username": {
                "id": "username",
                "type": "string"
            },
            "password": {
                "id": "password",
                "type": "string",
                "pattern": "^\$2b\$1[0-3]\$\S{53}$"
            },
            "firstName": {
                "id": "firstName",
                "type": "string",
                "pattern": "^(?:[\w\.]){3,60}"
            },
            "lastName": {
                "id": "lastName",
                "type": "string",
                "pattern": "^(?:[\w\.]){3,60}"
            },
            "experiments": {
                "id": "experiments",
                "type": "array",
                "items": {
                    "type": "string",
                    "pattern": JAM_ID_PATTERN
                },
                "uniqueItems": True
            }
        },
        "required": [
            "username",
            "password"
        ],
        "additionalProperties": False
    }
}

# default profile schema
PROFILE = {
    "type": "jsonschema",
    "schema": {
        "id": "profile",
        "type": "object",
        "properties": {
            "firstName": {
                "id": "firstName",
                "type": "string",
                "pattern": "^\w{3,64}",
            },
            "lastName": {
                "id": "lastName",
                "type": "string",
                "pattern": "^\w{3,64}",
            },
            "birthday": {
                "id": "birthday",
                "type": "string",
                "pattern": ISO_DATE_PATTERN
            },
            "account": {
                "id": "account",
                "type": "string",
                "pattern": JAM_ID_PATTERN
            },
            "sessions": {
                "id": "sessions",
                "type": "array",
                "items": {
                    "type": "string",
                    "pattern": JAM_ID_PATTERN,
                },
                "uniqueItems": True
            }
        },
        "required": [
            "firstName", "lastName", "birthday", "account"
        ],
        # "additionalProperties": False  // todo: re-enable
    }
}

CONFIG = {
    "type": "jsonschema",
    "schema": {
        "id": "config",
        "type": "object",
        "properties": {
            "profileSchema": {
                "id": "profileSchema",
                "type": "object"
            },
            "profilesMin": {
                "id": "profilesMin",
                "type": "integer"
            },
            "profilesMax": {
                "id": "profilesMin",
                "type": "integer"
            }
        },
        "additionalProperties": True
    }
}

EXPERIMENT = {
    "type": "jsonschema",
    "schema": {
        "id": "experiment",
        "type": "object",
        "properties": {
            "structure": {
                "id": "structure",
                "type": "array",
                "items": {
                    "type": "object"
                }
            },
            "title": {
                "id": "title",
                "type": "string"
            },
            "description": {
                "id": "description",
                "type": "string"
            },
            "active": {
                "id": "active",
                "type": "boolean"
            },
            "eligibilityCriteria": {
                "id": "eligibilityCriteria",
                "type": "object"
            }
        },
        "required": [
            "structure",
            "active"
        ],
        # "additionalProperties": False // TODO re-enable
    }
}

SESSION = {
    "type": "jsonschema",
    "schema": {
        "id": "session",
        "type": "object",
        "properties": {
            "profileId": {
                "id": "profileId",
                "type": "string"
            },
            "profileVersion": {
                "id": "profileVersion",
                "type": "string"
            },
            "experimentId": {
                "id": "experimentId",
                "type": "string"
            },
            "experimentVersion": {
                "id": "experimentVersion",
                "type": "string"
            },
            "parameters": {
                "id": "parameters",
                "type": "object"
            },
            "softwareVersion": {
                "id": "softwareVersion",
                "type": "string"  # TODO pattern? semver?
            },
            "data": {
                "id": "data",
                "type": "object"
            },
            "timestamp": {
                "id": "timestamp",
                "type": "string",
                "pattern": ISO_DATE_PATTERN
            }
        },
        "required": [
            "profileId", "profileVersion",
            "experimentId", "experimentVersion",
            "parameters", "softwareVersion",
            "data", "timestamp"
        ],
        "additionalProperties": False
    }
}

ACCOUNT = {
    "type": "jsonschema",
    "schema": {
        "id": "account",
        "type": "object",
        "properties": {
            "username": {  # TODO can this be an id?
                "id": "username",
                "type": "string",
                # "pattern": commonregex.email.pattern
            },
            "password": {
                "id": "password",
                "type": "string",
                "pattern": "^\$2b\$1[0-3]\$\S{53}$"
            },
        },
        "required": [
            "password"
        ],
        # "additionalProperties": False
    }
}

for schema in (ADMIN, CONFIG, EXPERIMENT, SESSION, ACCOUNT, PROFILE):
    with open('{}/../schemas/{}.json'.format(os.path.dirname(__file__), schema['schema']['id']), 'w') as fp:
        print('Made schema for {}'.format(schema['schema']['id']))
        json.dump(schema, fp, indent=4, sort_keys=True)

File: scripts/generate_schemas.js
/* globals require */
/* jshint quotmark: true */


var rimraf = require('rimraf');
var path = require('path');
var fs = require('fs');

// h/t: https://www.safaribooksonline.com/library/view/regular-expressions-cookbook/9781449327453/ch04s07.html
var ISO_DATE_PATTERN = '^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])(\\.[0-9]+)?(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?$';

// var URL_PATTERN = '(http|https)://[\\w-]+(\\.[\\w-]+)+([\\w.,@?^=%&amp;:/~+#-]*[\\w@?^=%&amp;/~+#-])?';

var JAM_ID_PATTERN = '\\w+';
var PROFILE_ID_PATTERN = '\\w+\\.\\w+';


var CONFIG = {
    "schema": {
        "id": "config",
        "type": "object",
        "properties": {
            "profilesMin": {
                "id": "profilesMin",
                "type": "integer"
            },
            "profilesMax": {
                "id": "profilesMin",
                "type": "integer"
            }
        },
        "additionalProperties": true
    },
    "type": "jsonschema"
};

var EXPERIMENT = {
    "schema": {
        "id": "experiment",
        "type": "object",
        "properties": {
            "title": {
                "id": "title",
                "type": "string"
            },
            "description": {
                "id": "description",
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "state": {
                "id": "state",
                "type": "string",
                "enum": [
                    "Draft",
                    "Active",
                    "Archived",
                    "Deleted"
                ]
            },
            "beginDate": {
                "id": "beginDate",
                "format": "date-time",
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "endDate": {
                "id": "endDate",
                "format": "date-time",
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "structure": {
                "id": "structure",
                "type": "object",
                "properties": {
                    "frames": {
                        "type": "object"
                    },
                    "sequence": {
                        "type": "array",
                        "items": {
                            "type": "string"
                        }
                    }
                }
            },
            "eligibilityCriteria": {
                "id": "eligibilityCriteria",
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "displayFullscreen": {
                "type": "boolean",
                "default": false
            },
            "exitUrl": {
                "type": "string",
                "default": "/"
            },
            "duration": {
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "purpose": {
                "$oneOf": [
                    "string",
                    null
                ]
            }
        },
        "required": [
            "structure",
            "state"
        ]
    },
    "type": "jsonschema"
};


var SESSION = {
    "schema": {
        "id": "sessiontest0",  // Script creates one particular session collection associated with one single experiment
        "type": "object",
        "properties": {
            "completed": {
                "type": "boolean"
            },
            "profileId": {
                "type": "string",
                "pattern": PROFILE_ID_PATTERN
            },
            "experimentId": {  // TODO: In the new collection-per-experiment sessions model, this field may be redundant
                "type": "string",
                "pattern": JAM_ID_PATTERN
            },
            "experimentVersion": {
                "type": "string"
            },
            "sequence": {
                "type": "array",
                "items": {
                    "type": "string"
                }
            },
            "expData": {
                "type": "object"
            },
            "feedback": {
                "$oneOf": [{
                    "type": "string"
                }, null]
            },
            "hasReadFeedback": {
                "$oneOf": [{
                    "type": "boolean"
                }, null]
            },
            "earlyExit": {
                "$oneOf": [{
                    "type": "object"
                }, null],
                "properties": {
                    "reason": {
                        "$oneOf": [{
                            "type": "string"
                        }, null]

                    },
                    "privacy": {
                        "type": "string"
                    }
                }
            }
        },
        "required": [
            "profileId",
            "experimentId",
            "experimentVersion",
            "completed",
            "sequence",
            "expData"
        ]
        // "additionalProperties": false
    },
    "type": "jsonschema"
};

var ACCOUNT = {
    "schema": {
        "id": "account",
        "type": "object",
        "properties": {
            "username": {  // TODO can this be an id?
                "type": "string"
                // # "pattern": commonregex.email.pattern
            },
            "name": {
                "$oneOf": [
                    "string",
                    null
                ]
            },
            "password": {
                "id": "password",
                "type": "string",
                "pattern": "^\\$2b\\$1[0-3]\\$\\S{53}$"
            },
            "profiles": {
                "type": "array",  // Could also do this as an object, as long as keys were guaranteed unique
                "items": {
                    "type": "object",
                    "properties": {
                        "profileId": {
                            "type": "string",
                            "pattern": PROFILE_ID_PATTERN
                        },
                        "firstName": {
                            "type": "string",
                            "pattern": "^\\w{3,64}"
                        },
                        "birthday": {
                            "type": "string",
                            "pattern": ISO_DATE_PATTERN
                        },
                        "deleted": {
                            "type": "boolean",
                            "default": false
                        }
                    },
                    "required": ["firstName", "birthday"]
                }
            }
        },
        "required": ["username", "password"]
        // "additionalProperties": false
    },
    "type": "jsonschema"
};

// TODO RE-add account
module.exports = function main() {
    var base = path.dirname(__filename);
    rimraf.sync(`${base}/../schemas/*`);
    [EXPERIMENT, SESSION, ACCOUNT].forEach(function(schema) {
        var schemaData = JSON.stringify(schema, null, 4);
        var filename = schema.schema.id;
        fs.writeFile(`${base}/../schemas/${filename}.json`, schemaData);
    });
};
